### PR TITLE
Fix for output type and null subscriptions collection in ResourceGraph

### PR DESCRIPTION
**Pull request by fenwickt**
## Description

- Fixed a bug with null subscriptions collection trying to be casted to list
- Copied changes from this PR of fixing output https://github.com/Azure/azure-powershell/pull/15125 and updated unit tests
- Added ChangeLog description

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/ResourceGraph/ResourceGraph/Az.ResourceGraph.psd1 [hunk 1]
```diff
@@ -70,7 +70,7 @@ FormatsToProcess = 'ResourceGraph.format.ps1xml',
                'ResourceGraph.Autorest\Az.ResourceGraph.format.ps1xml'
 
 # Modules to import as nested modules of the module specified in RootModule/ModuleToProcess
-NestedModules = @('Az.ResourceGraph.psm1', 
+NestedModules = @('Microsoft.Azure.PowerShell.Cmdlets.ResourceGraph.dll',
```
Comment (fenwickt) [inline src/ResourceGraph/ResourceGraph/Az.ResourceGraph.psd1:73]: Please let me know if I should revert this change. I wasn't able to run tests locally without it.

**Comment by dramuy** [comment]
@VeryEarly is added to the review. #Closed

File: src/ResourceGraph/ResourceGraph/ChangeLog.md [hunk 2]
```diff
@@ -18,6 +18,7 @@
         - Additional information about change #1
 -->
 ## Upcoming Release
+* Fixed the output print issue for `Search-AzGraph`. Fixed the issue when Search-AzGraph fails  when no subscriptions are stored in the context.
```
Comment (dramuy) [inline src/ResourceGraph/ResourceGraph/ChangeLog.md:21]: >   [](http://example.com/codeflow?start=94&length=2)

nit: multiple spaces

File: src/ResourceGraph/ResourceGraph/Models/PSResourceGraphResponse.cs [hunk 3]
```diff
@@ -14,17 +14,49 @@
...
 
-    public class PSResourceGraphResponse
+    public class PSResourceGraphResponse<PSObject> : IList<PSObject>
     {
         [Ps1Xml(Target = ViewControl.List)]
         public string SkipToken { get; set; }
 
         [Ps1Xml(Target = ViewControl.List)]
         public IList<PSObject> Data { get; set; }
+        public PSObject this[int index] { get => Data[index]; set => Data[index] = value; }
```
Comment (dramuy) [inline src/ResourceGraph/ResourceGraph/Models/PSResourceGraphResponse.cs:28]: >public PSObject this[int index] { [](http://example.com/codeflow?start=8&length=33)

nit: add a new line here.

File: src/ResourceGraph/ResourceGraph/Models/PSResourceGraphResponse.cs [hunk 4]
```diff
@@ -14,17 +14,49 @@
...
 {
+    using System.Collections;
     using System.Collections.Generic;
-    using System.Management.Automation;
     using Microsoft.WindowsAzure.Commands.Common.Attributes;
 
-    public class PSResourceGraphResponse
+    public class PSResourceGraphResponse<PSObject> : IList<PSObject>
     {
         [Ps1Xml(Target = ViewControl.List)]
```
Comment (dramuy) [inline src/ResourceGraph/ResourceGraph/Models/PSResourceGraphResponse.cs:23]: >[Ps1Xml(Target = ViewControl.List)] [](http://example.com/codeflow?start=8&length=35)

are these Ps1xml attributes with List control still accurate and applies?
and any ps1xml file changes needed?

File: src/ResourceGraph/ResourceGraph/help/Search-AzGraph.md [hunk 5]
```diff
@@ -74,7 +75,7 @@ id         : /subscriptions/1ef51df4-f8a9-4b69-9919-1ef51df4eff6/resourceGroups/
 name       : egtopic-2
```
Comment (dramuy) [inline src/ResourceGraph/ResourceGraph/help/Search-AzGraph.md:75]: please validate that the cmdlet results are in this format as well with the new changes.

File: src/ResourceGraph/ResourceGraph/Models/PSResourceGraphResponse.cs [hunk 6]
```diff
@@ -14,17 +14,49 @@
 
 namespace Microsoft.Azure.Commands.ResourceGraph.Models
 {
+    using System.Collections;
     using System.Collections.Generic;
-    using System.Management.Automation;
     using Microsoft.WindowsAzure.Commands.Common.Attributes;
 
-    public class PSResourceGraphResponse
+    public class PSResourceGraphResponse<PSObject> : IList<PSObject>
```
Comment (dramuy) [inline src/ResourceGraph/ResourceGraph/Models/PSResourceGraphResponse.cs:21]: >PSResourceGraphResponse<PSObject> : IList<PSObject> [](http://example.com/codeflow?start=17&length=51)

@VeryEarly thank you very much for all the help here addressing our cmdlets compatibility issues and formatter concers. You Rock !!!:thumbsup: :rocket:

File: src/ResourceGraph/ResourceGraph/Models/PSResourceGraphResponse.cs (hunk 4 above)
Comment (fenwickt) [reply src/ResourceGraph/ResourceGraph/Models/PSResourceGraphResponse.cs:23]: Good point for the ps1xml change, thank you! Updated it.
Technically List keyword is not changing anything anymore. Though I still left the List attribute since the default value is Table and List makes more sense to me for output of data and skip token.

File: src/ResourceGraph/ResourceGraph/ResourceGraph.format.ps1xml [hunk 7]
```diff
@@ -1,30 +1,25 @@
...
         <ListEntries>
           <ListEntry>
             <ListItems>
               <ListItem>
                 <PropertyName>SkipToken</PropertyName>
-                <ItemSelectionCondition>
-                  <ScriptBlock>
-                    $_.SkipToken -ne $null
-                  </ScriptBlock>
-                </ItemSelectionCondition>
```
Comment (dramuy) [inline src/ResourceGraph/ResourceGraph/ResourceGraph.format.ps1xml:19]: is this removal based on generated ps1xml file.

File: src/ResourceGraph/ResourceGraph/help/Search-AzGraph.md [hunk 8]
```diff
@@ -64,34 +66,43 @@ A complex query on resources featuring field selection, filtering and summarizin
...
 
 id         : /subscriptions/1ef51df4-f8a9-4b69-9919-1ef51df4eff6/resourceGroups/Service-INT-a/providers/Microsoft.EventGrid/topics/egtopic-1
 name       : egtopic-1
 type       : microsoft.eventgrid/topics
 location   : westus2
 tags       :
+ResourceId : /subscriptions/1ef51df4-f8a9-4b69-9919-1ef51df4eff6/resourceGroups/Service-INT-a/providers/Microsoft.EventGrid/topics/egtopic-1
 ```
 
 A query scoped to the management group that allows the query to succeed with partial scope result if MyManagementGroupId has more than N subscriptions underneath.
```
Comment (dramuy) [inline src/ResourceGraph/ResourceGraph/help/Search-AzGraph.md:108]: >more than N [](http://example.com/codeflow?start=125&length=11)

this statement feels keeping it open question to end customers thinking about what is the N means for my tenant. N => 1000 or 5000.

Should we add information on - 

if the given management group exceeds the limit of number of allowed subscriptions while querying for resources. Also provide information that while querying for container resources we don't have any limits.

**Review by VeryEarly — APPROVED** [review]

File: src/ResourceGraph/ResourceGraph/ResourceGraph.format.ps1xml (hunk 7 above)
Comment (fenwickt) [reply src/ResourceGraph/ResourceGraph/ResourceGraph.format.ps1xml:19]: Yes, it is
